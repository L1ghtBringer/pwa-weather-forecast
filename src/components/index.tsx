export { default as Calendar } from './Calendar'
export { default as DatePicker } from './DatePicker'
export { default as Select } from './Select'
export { default as FillFields } from './FillFields'
export { default as WeatherCard } from './WeatherCard'
export { default as Carousel } from './Carousel'
export { default as Snackbar } from './Snackbar'
