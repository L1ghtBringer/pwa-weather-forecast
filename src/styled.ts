import baseStyled, { ThemedStyledInterface } from 'styled-components'

export const theme = {
  media: {
    laptop: '(max-width: 1366px)',
    tablet: '(max-width: 805px)',
    tabletS: '(max-width: 700px)',
    phone: '(max-width: 520px)'
  },
  color: {
    accent: '#373AF5',
    secondary: {
      active: '#8083A4',
      hover: '#8083A499'
    },
    baseStong: '#2C2D76',
    critic: '#E5596D',
    baseWeak: '#FFFFFF'
  },
  font: 'Ubuntu, sans-serif'
}

export type Theme = typeof theme

export const styled = baseStyled as ThemedStyledInterface<Theme>
