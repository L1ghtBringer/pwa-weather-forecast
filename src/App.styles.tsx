import styled from 'styled-components'

export const AppWrapper = styled.div`
  width: 100%;
  min-height: 100vh;
  height: auto;
  position: relative;
`

export const LoveMercury = styled.p`
  color: #FFFFFF;
  opacity: 0.6;
  font-family: Ubuntu, sans-serif;
  font-size: 14px;
  position: absolute;
  left: calc(50% - 140px);
  bottom: 16px;
`
