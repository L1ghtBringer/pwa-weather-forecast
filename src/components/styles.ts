import styled from 'styled-components'

export const Input = styled.input`
  width: 100%;
  font-family: ${props => props.theme.font};
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  padding: 10px 16px;
  background: rgba(128, 131, 164, 0.06);
  border: 2px solid rgba(128, 131, 164, 0.2);
  border-radius: 8px;
  outline: none;
  color: #2C2D76;
  :hover{
    border: 2px solid #8083A4;
  }
  :focus{
    border: 2px solid #373AF5;
  }
  -webkit-transform: translate3d(0,0,0);
  transition: border 0.4s ease-out;
  cursor: pointer;
`

export const Container = styled.div`
  max-width: 252px;
  width: 100%;
  position: relative;
  img {
    position: absolute;
    pointer-events: none;
  }
  @media ${props => props.theme.media.phone} {
    max-width: 100%;
  }
`
