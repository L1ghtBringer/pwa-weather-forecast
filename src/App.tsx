import React from 'react'
import Forecast from 'pages/Forecast/Forecast'
import { AppWrapper, LoveMercury } from './App.styles'

const App: React.FC = () => {
  return <AppWrapper>
    <Forecast/>
    <LoveMercury>C ЛЮБОВЬЮ ОТ MERCURY DEVELOPMENT</LoveMercury>
  </AppWrapper>
}

export default App
