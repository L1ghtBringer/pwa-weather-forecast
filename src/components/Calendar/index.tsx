import React, { useState } from 'react'
import { ICalendarProps } from './interfaces'
import { Button, CalendarCard, Footer, Header, Table } from './calendar.styles'
import { getMonthData, areEqual, qt } from './calendar'
import { locale } from './locale/en_GB'

const defaultProps: ICalendarProps = {
  locale,
  years: [2020, 2021, 2022, 2023, 2024, 2025],
  onChange: (date: Date) => {}
}

interface IState {
  date: Date
  currentDate: Date
  selectedDate: Date | null
}

const Calendar: React.FC<ICalendarProps> = ({
  years,
  locale,
  selectedDate,
  blockDay,
  onChange,
  style
}) => {
  const [state, setState] = useState<IState>({
    date: selectedDate ?? new Date(),
    currentDate: new Date(),
    selectedDate: selectedDate ?? null
  })

  const year: number = state.date.getFullYear()
  const month: number = state.date.getMonth()

  const monthDate = getMonthData(year, month, locale ? locale.startWeek : 'Su')

  const handlePrevMonthButtonClick = (): void => {
    setState((prevState) => ({
      ...prevState,
      date: new Date(year, month - 1)
    }))
  }

  const handleNextMonthButtonClick = (): void => {
    setState((prevState) => ({
      ...prevState,
      date: new Date(year, month + 1)
    }))
  }

  const handleDayClick = (date: Date): void => {
    setState((prevState) => ({ ...prevState, date, selectedDate: date }))
    onChange!(date)
  }

  const setTodayDay = (): void => {
    if (!disabledDay(new Date()) && onChange) {
      const today = new Date()
      setState((prevState) => ({
        ...prevState,
        date: today,
        selectedDate: today
      }))
      onChange(new Date(today.getFullYear(), today.getMonth(), today.getDate()))
    }
  }

  const disabledDay = (date: Date): boolean => {
    if (!blockDay) return false
    if (qt(date, blockDay.afterDay) || !qt(date, blockDay.beforeDay)) {
      return true
    }
    return false
  }

  return (
    <CalendarCard styleUser={String(style)}>
      <Header>
        <Button
          padding={'.5em'}
          onClick={handlePrevMonthButtonClick}
          disabled={new Date(year, month - 1).getFullYear() < years![0]}
        >
          {'<'}
        </Button>
        <span>{locale!.monthNames[month]}</span>
        <span>{year}</span>
        <Button
          padding={'.5em'}
          onClick={handleNextMonthButtonClick}
          disabled={
            new Date(year, month + 1).getFullYear() > years![years!.length - 1]
          }
        >
          {'>'}
        </Button>
      </Header>
      <Table>
        <thead>
          <tr>
            {locale!.weekDayNames.map((name) => (
              <th key={name}>{name}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {monthDate.map((week, index) => (
            <tr key={index}>
              {week.map((date, index) => (
                <td key={index}>
                  <Button
                    isDate={true}
                    today={areEqual(date, state.currentDate)}
                    selected={areEqual(date, state.selectedDate!)}
                    onClick={() => handleDayClick(date!)}
                    thisMonth={date?.getMonth() === state.date.getMonth()}
                    disabled={disabledDay(date!)}
                  >
                    {date!.getDate()}
                  </Button>
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </Table>
      <Footer>
        <Button disabled={disabledDay(new Date())} onClick={setTodayDay}>
          {locale!.today}
        </Button>
      </Footer>
    </CalendarCard>
  )
}

Calendar.defaultProps = defaultProps

export default Calendar
