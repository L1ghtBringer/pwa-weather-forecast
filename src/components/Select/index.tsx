import React, { useState } from 'react'
import { ISelectProps, IOption } from './interfaces'
import { useClickOutside } from '../../hooks/clickOutside'
import { SelectContainer, OptionContainer, Option } from './select.styles'
import { Input } from '../styles'
import arrow from './icons/arrow.svg'

const defaultProps: ISelectProps = {
  placeholder: 'Select data',
  onChange: () => {},
  value: {
    latitude: null,
    longitude: null
  }
}

const Select: React.FC<ISelectProps> = ({
  placeholder,
  options,
  onChange,
  value
}) => {
  const [text, setText] = useState<string>('')
  const [open, setOpen] = useState<boolean>(false)
  const myRef = useClickOutside<HTMLDivElement>(handleClickOutside)

  function handleClickOutside (event: Event): void {
    setOpen(false)
  }

  const handleClickOption = (option: IOption): void => {
    setOpen(false)
    if (
      !(
        value?.latitude === option.value.latitude &&
        value?.longitude === option.value.longitude
      )
    ) {
      onChange!(option.value)
      setText(option.label)
    }
  }

  return (
    <SelectContainer ref={myRef} visible={open}>
      <Input
        placeholder={placeholder}
        value={text}
        onClick={() => setOpen((prevState) => !prevState)}
        readOnly
      />
      <img src={arrow} alt="icon-arrow" />
      <OptionContainer visible={open}>
        <div>
          {options!.map((element) => (
            <Option
              key={element.label}
              onClick={() => handleClickOption(element)}
              visible={open}
            >
              {element.label}
            </Option>
          ))}
        </div>
      </OptionContainer>
    </SelectContainer>
  )
}

Select.defaultProps = defaultProps

export default Select
