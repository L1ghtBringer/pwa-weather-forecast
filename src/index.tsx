import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import { createGlobalStyle, ThemeProvider } from 'styled-components'
import backgroundDown from 'assets/backgroundDown.svg'
import backgroundDownLow from 'assets/backgroundDownLow.svg'
import backgroundUp from 'assets/backgroundUp.svg'
import backgroundUpLow from 'assets/backgroundUpLow.svg'
import { theme } from './styled'
import './font.css'

const Global = createGlobalStyle`
*,
::after,
::before{
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
body {
  overflow-x:hidden;
  margin-right: calc(-1 * (100vw - 100%));
  background: url(${backgroundUp}) top center no-repeat, url(${backgroundDown}) bottom center no-repeat #373af5;;
  background-size: 100%, 100%, cover;
  @media ${theme.media.phone}{
    background: url(${backgroundUpLow}) top center no-repeat, url(${backgroundDownLow}) bottom center no-repeat #373af5;
    background-size: 100%, 100%, cover;
  }
}
`

ReactDOM.render(
  <React.StrictMode>
    <ThemeProvider theme={theme}>
      <Global />
      <App />
    </ThemeProvider>
  </React.StrictMode>,
  document.getElementById('root')
)
