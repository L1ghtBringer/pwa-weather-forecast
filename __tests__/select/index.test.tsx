import React from 'react'
import { shallow, mount, configure } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import Select from '../../src/components/Select'
import { ThemeProvider } from 'styled-components'
import { theme } from '../../src/styled'
import {
  Option,
  SelectContainer,
} from '../../src/components/Select/select.styles'
import { Input } from '../../src/components/styles'
import { act } from 'react-dom/test-utils'
import { createDocumentListenersMock } from '../createDocumentListenersMock'

configure({ adapter: new Adapter() })

const shallowWithTheme = (children: any) =>
  shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

const mountWithTheme = (children: any) =>
  mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

describe('Select component test', () => {
  it('Should match the snapshot', () => {
    const select = shallowWithTheme(<Select options={cities} />)
    expect(select.html()).toMatchSnapshot()
    expect(select.exists()).toBe(true)
  })

  it('Open window', () => {
    const select = shallow(<Select options={cities} />)
    expect(select.props().visible).toBe(false)
    select.find(Input).simulate('click')
    expect(select.props().visible).toBe(true)
  })

  it('Test select + reselect', () => {
    const select = shallow(<Select options={cities} />)
    const option = select.find(Option).at(0).text()
    expect(select.find(Input).props().value).toBe('')
    select.find(Option).at(0).simulate('click')
    expect(select.find(Input).props().value).toBe(option)
  })

  it('Test reselect', () => {
    const select = shallow(<Select options={cities} />)
    const option = select.find(Option).at(0).text()
    expect(select.find(Input).props().value).toBe('')
    select.find(Option).at(0).simulate('click')
    expect(select.find(Input).props().value).toBe(option)

    const optionTwo = select.find(Option).at(1).text()
    select.find(Option).at(1).simulate('click')
    expect(select.find(Input).props().value).toBe(optionTwo)
  })

  it('Change test', () => {
    const onChange = jest.fn()
    const select = shallow(<Select options={cities} onChange={onChange} />)
    select.find(Option).at(0).simulate('click')
    expect(onChange).toHaveBeenCalledTimes(1)
  })

  it('One change test', () => {
    const value = {
      latitude: 55.796127,
      longitude: 49.106405,
    }
    const onChange = jest.fn()
    const select = shallow(
      <Select options={cities} value={value} onChange={onChange} />
    )
    select.find(Option).at(0).simulate('click')
    select.find(Option).at(0).simulate('click')
    expect(onChange).toHaveBeenCalledTimes(0)
  })

  it('Click outside', () => {
    const onChange = jest.fn()
    const fireEvent = createDocumentListenersMock();
    const select = mountWithTheme(<Select options={cities} onChange={onChange} />)
    select.find(Input).at(0).simulate('click')
    expect(select.find(SelectContainer).props().visible).toBe(true)
    act(()=>{
        fireEvent.mouseDown(document.body);
    })
    select.find(Select).at(0).simulate('mousedown')
    expect(select.find(SelectContainer).props().visible).toBe(false)
  })
})

const cities = [
  {
    value: {
      latitude: 55.796127,
      longitude: 49.106405,
    },
    label: 'Kazan',
  },
  {
    value: {
      latitude: 45.03547,
      longitude: 38.975313,
    },
    label: 'Krasnodar',
  },
  {
    value: {
      latitude: 53.195873,
      longitude: 50.100193,
    },
    label: 'Samara',
  },
  {
    value: {
      latitude: 51.533557,
      longitude: 46.034257,
    },
    label: 'Saratov',
  },
  {
    value: {
      latitude: 53.507836,
      longitude: 49.420393,
    },
    label: 'Tolyatti',
  },
]
