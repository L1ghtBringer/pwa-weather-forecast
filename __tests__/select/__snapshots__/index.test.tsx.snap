// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component test Should match the snapshot 1`] = `"<div class=\\"sc-gtsrHT sc-dlnjwi fAOKgM cZXYwK\\"><input placeholder=\\"Select data\\" value=\\"\\" readonly=\\"\\" class=\\"sc-bdnxRM dcrmCf\\"/><img src=\\"[object Object]\\" alt=\\"icon-arrow\\"/><div class=\\"sc-hKFxyN hnSJtu\\"><div><div class=\\"sc-bdnxRM dcrmCf sc-eCApnc bQcmnQ\\">Kazan</div><div class=\\"sc-bdnxRM dcrmCf sc-eCApnc bQcmnQ\\">Krasnodar</div><div class=\\"sc-bdnxRM dcrmCf sc-eCApnc bQcmnQ\\">Samara</div><div class=\\"sc-bdnxRM dcrmCf sc-eCApnc bQcmnQ\\">Saratov</div><div class=\\"sc-bdnxRM dcrmCf sc-eCApnc bQcmnQ\\">Tolyatti</div></div></div></div>"`;
